Some fixes to the Genie* classes

penev92 opened the pull request:
Also AoE :D

To test put `GRAPHICS.DRS` and `50500.bin` in `\OpenRA\Content\aoe\`, launch OpenRA, and use the map editor to see previews of some actors (don't try to save a map with actors on it, because they are not properly set up - they work only in the map editor).

penev92 committed:
Add GenieSlpHeader class

penev92 committed:
Some fixes to GenieSlpLoader

penev92 committed:
Add a PaletteFromJascFile trait

penev92 committed:
Add a placeholder Age of Empires mod

phrohdoh commented on OpenRA.Mods.Common/SpriteLoaders/GenieSlpLoader.cs:
```diff
@@ -33,20 +49,30 @@ public GenieSlpFrame(Stream stream, GenieSlpFrameHeader header)
 				var rowData = new List<byte>();
 				for (var x = 0; x < header.Width; x++)
-					rowData[x] = DEFAULT_INDEX;
+					rowData.Add(DEFAULT_INDEX);
```
This fills rows with the `DEFAULT_INDEX` which is not desired.
We only want to define data that actually exists.

phrohdoh commented on OpenRA.Mods.Common/SpriteLoaders/GenieSlpLoader.cs:
```diff
@@ -54,14 +80,14 @@ static byte[] ReadRowCommands(Stream stream, GenieSlpFrameHeader header)
 
-				var opcode = 0xf;
-				var rowPosX = 0;
+				int opcode;
```
This is now `0` instead of `0xf`.

penev92 replied:
This gets set 5 lines down, before it's ever used, so you were setting it twice.

phrohdoh commented on OpenRA.Mods.Common/SpriteLoaders/GenieSlpLoader.cs:
```diff
@@ -54,14 +80,14 @@ static byte[] ReadRowCommands(Stream stream, GenieSlpFrameHeader header)
 					var currByte = GetSingleByte(stream);
 
-					opcode = currByte & 0x0f;
+					opcode = currByte & 15;
```
Was there any reason for changing this to integer literal from hex?

penev92 replied:
Easier readability, and they are also mixed throughout the code.

LavenderMoon replied:
Generally, Hex is easier to understand when using logical and/or. Of course, bitfields would be best, but I don't think they're usable in C#.

phrohdoh commented on OpenRA.Mods.Common/SpriteLoaders/GenieSlpLoader.cs:
```diff
@@ -89,11 +115,11 @@ static byte[] ReadRowCommands(Stream stream, GenieSlpFrameHeader header)
 
 					case 0x06: // Copy & Transform
-						for (uint i = 0, cmdLen = GetTopNibblePlusNext(currByte, stream); i < cmdLen; i++)
+						for (uint i = 0, cmdLen = GetTopNibbleOrNext(currByte, stream); i < cmdLen; i++)
```
What's the justification for this change?
Which docs did you reference?

penev92 replied:
Had to change it because it crashed.
Also confirmed the change with [this](https://github.com/Phrohdoh/aoe-parser-cs/blob/master/AgeSlp/SlpFrameRow.cs#L120).

phrohdoh commented on OpenRA.Mods.Common/SpriteLoaders/GenieSlpLoader.cs:
```diff
@@ -210,33 +236,32 @@ static bool IsGenieSlp(Stream stream)
 		{
 			var pos = stream.Position;
 			var test = stream.ReadASCII(32);
+			stream.Position = pos;
```
I was thinking about this fix on the way home, good call.

penev92 committed:
Some more fixes

penev92 committed:
Enable SLP preview in the AssetBrowser

penev92 commented:
Added a `Some more fixes to GenieSlpLoader` commit. Kept it separate so you can see what changed easier. You should squash those 2 commits into your original one.
Also added a "fix" for the AssetBrowser. It now displays SLPs.

P.S.: Also greatly reduced the amount of YAML in `defaults.yaml`, `player.yaml` and `world.yaml`.

phrohdoh commented:
Awesome work, thanks a bunch. :tada:
